### DISCUSSION TITLE
Generated SQL query has two repeated joins, how to avoid it?

### DISCUSSION THREAD

**Discussion by valeriob23** [sqlalchemy/sqlalchemy#6877] [Usage Questions] [answered]
Hello everyone,
I'm migrating to the new SQLAlchemy syntax and I'm facing some issues generating the SQL query I need.
In detail: 
- I'm using the imperative mapping to map tables with my domain models
- I've a relationship between two tables, let's say _Address_ and _User_, and I need a simple query which joins them and filters by a given _User_ field
- I need a _joined loading_ due to performance reasons (therefore I use `lazy='joined'`)

You can find all details related to domain models, tables and relationship below. I took them from documentation, modifying them just to show you an example and reflect my real case:

```python
from __future__ import annotations

from dataclasses import dataclass
from dataclasses import field
from typing import List

from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.orm import registry
from sqlalchemy.orm import relationship

mapper_registry = registry()

@dataclass
class User:
    id: int = field(init=False)
    name: str = None
    fullname: str = None
    nickname: str = None

@dataclass
class Address:
    user: User
    id: int = field(init=False)
    email_address: str = None

metadata = MetaData()

user = Table(
    'user',
    metadata,
    Column('id', Integer, primary_key=True),
    Column('name', String(50)),
    Column('fullname', String(50)),
    Column('nickname', String(12)),
)

address = Table(
    'address',
    metadata,
    Column('id', Integer, primary_key=True),
    Column('user_id', Integer, ForeignKey('user.id')),
    Column('email_address', String(50)),
)

mapper_registry.map_imperatively(User, user)

mapper_registry.map_imperatively(Address, address, properties={
    'user': relationship(User, lazy='joined', innerjoin=True)
})
```

I've tried different approach to build my query, but no one satisfies me.

**Approach 1:** 
```python
stmt = select(Address).where(User.name.like("%whatever%"))
```
It generates a query with both tables in FROM clause, generating an unexpected result (I'll omit columns in SELECT clause):
```sql
SELECT ... 
FROM "user", address JOIN "user" AS user_1 ON user_1.id = address.user_id 
WHERE "user".name LIKE :name_1
```

**Approach 2:** 
To avoid the situation above, I tried to specify the join:
```python
stmt = select(Address).join(User).where(User.name.like("%whatever%"))
```
However, it generates a query with two repeated joins. The execution plan shows an additional "cost", therefore I'd avoid this approach:
```sql
SELECT ... 
FROM address JOIN "user" ON "user".id = address.user_id JOIN "user" AS user_1 ON user_1.id = address.user_id 
WHERE "user".name LIKE :name_1
```

_Please note:_ Without `lazy='joined'`, the Approach 2 would generate the correct query. However, as mentioned before, I need that configuration due to perfomance reasons.

What would you suggest? How can I get the correct query while maintaining this configurations? 
Thanks!

### Versions
- OS: MacOS Catalina 10.15.7
- Python: 3.9
- SQLAlchemy: 1.4.22
- Database: Postgres
- DBAPI: 2.0 (async engine)

**Comment by CaselIT** [comment, marked as answer]
Hi,

First of all, all the emoji reactions are not really needed for a discussion to be replied to

The behaviour of joineload is documented at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#joined-eager-loading

> Approach 1:

This works as explected, No join is specified in `select(Address).where(User.name.like("%whatever%"))` so two froms are rendered

> Approach 2:

Again as expected. .See this [paragraph](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#the-zen-of-joined-eager-loading) in the above documentation

> What would you suggest? How can I get the correct query while maintaining this configurations?

Off the top of my head I think you have two options:
- use the same query as approach 2 but use instead the [`contains_eager`](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#using-contains-eager-to-load-a-custom-filtered-collection-result) options to indicate that the query already contains the `User` columns
- do not add a join, similar to approach a, and instead use [`has`](https://docs.sqlalchemy.org/en/14/orm/internals.html#sqlalchemy.orm.RelationshipProperty.Comparator.has) in the where. Something like `select(Address).where(Address.user.has(User.name.like("%whatever%")))`

Note that this is not a new behaviour of the 2.0 syntax. It's the same when using query

**Reply by valeriob23** [reply]
Hi,
thanks for your feedback. I think `contains_eager` is the best solution for my case, as it produces exactly the SQL query I need.
I really appreciate your support, thanks again! 

UPDATE: I post here the final statement for those who need it:
`select(Address).join(User).options(contains_eager(Address.user)).where(User.name.like("%whatever%"))`
